readthedocs/core/permissions.py

```diff
@@ -1,18 +1,86 @@
-# -*- coding: utf-8 -*-
-
 """Objects for User permission checks."""
+from django.conf import settings
 from django.contrib.auth.models import User
 from django.db.models import Q
 
 from readthedocs.core.utils.extend import SettingsOverrideObject
+from readthedocs.organizations.constants import ADMIN_ACCESS, READ_ONLY_ACCESS
 
 
 class AdminPermissionBase:
 
     @classmethod
-    def admins(cls, obj):
+    def projects(cls, user, admin=False, member=False):
+        """
+        Return all the projects the user has access to as ``admin`` or ``member``.
+
+        If `RTD_ALLOW_ORGANIZATIONS` is enabled
+        this function takes SSO into consideration, it includes:
+
+        - Projects where the user has access to via SSO.
+        - Projects where the user has access via Team,
+          if SSO is not enabled for the organization of that team.
+
+        .. note::
+
+           SSO is taken into consideration,
+           but isn't implemented in .org yet.
+
+        :param user: user object to filter projects
+        :type user: django.contrib.admin.models.User
+        :param bool admin: include projects where the user has admin access to the project
+        :param bool member: include projects where the user has read access to the project
+        """
         from readthedocs.projects.models import Project
+        projects = Project.objects.none()
+        if not user or not user.is_authenticated:
+            return projects
+
+        if not settings.RTD_ALLOW_ORGANIZATIONS:
+            # All users are admin and member of a project
+            # when we aren't using organizations.
+            return user.projects.all()
+
+        if admin:
+            # Project Team Admin
+            admin_teams = user.teams.filter(access=ADMIN_ACCESS)
+            for team in admin_teams:
+                if not cls._has_sso_enabled(team.organization):
+                    projects |= team.projects.all()
+
+            # Org Admin
+            for org in user.owner_organizations.all():
+                if not cls._has_sso_enabled(org):
+                    # Do not grant admin access on projects for owners if the
+                    # organization has SSO enabled with Authorization on the provider.
+                    projects |= org.projects.all()
+
+            projects |= cls._get_projects_for_sso_user(user, admin=True)
+
+        if member:
+            # Project Team Member
+            member_teams = user.teams.filter(access=READ_ONLY_ACCESS)
+            for team in member_teams:
+                if not cls._has_sso_enabled(team.organization):
+                    projects |= team.projects.all()
+
+            projects |= cls._get_projects_for_sso_user(user, admin=False)
+
+        return projects
+
+    @classmethod
+    def _has_sso_enabled(cls, organization):
```

Comment:
This method should be public (without the `_` prefix), since it's used in different places outside these querysets.

Reply:
Also, it should receive `provider=` to check if has VCS or Google SSO

Reply:
This is only used here, it acts as a proxy for the `has_sso_enabled` function from .com, is not the same.

Reply:
I want to reduce the overrides for the same method/function name. With this change, we have 3 `has_sso_enabled` that do different things. There are 2 that call the "original" one with different arguments; which is confusing and hard to follow. I prefer if we don't add these 2 extras and always use the same one.

Reply:
what about making provider optional `None` and default to ALL_AUTH? but looks like calling has_sso_enabled with None is valid... so what about naming this `has_allauth_sso_enabled`?

Reply:
I'd say no to this proposal. These are my reasons:

- calling with `None` checks for "any type of SSO enabled"
- making the call explicit where it's used is better to understand what's happening in that code and easier to read
- we currently have only two (ALL_AUTH and GOOGLE), but this may change in the near future
  - SAML
  - OKTA
  - split ALL_AUTH into GITHUB, GITLAB, BITBUCKET; since I found cases where this will be useful